File: src/behavioral/visitor.java
@Data
public abstract class Client {
    private final String name;
    private final String address;
    private final String number;
}

@Getter
public class Resident extends Client {
    private final String insuranceClass;

    public Resident(String name, String address, String number, String insuranceClass) {
        super(name, address, number);
        this.insuranceClass = insuranceClass;
    }
}

@Getter
public class Company extends Client {
    private final int nbrOfEmployees;

    public Company(String name, String address, String number, int nbrOfEmployees) {
        super(name, address, number);
        this.nbrOfEmployees = nbrOfEmployees;
    }
}

@Getter
public class Bank extends Client {
    private final int branchesInsured;

    public Bank(String name, String address, String number, int branchesInsured) {
        super(name, address, number);
        this.branchesInsured = branchesInsured;
    }
}

@Getter
public class Restaurant extends Client {
    private final boolean availableAbroad;

    public Restaurant(String name, String address, String number, boolean availableAbroad) {
        super(name, address, number);
        this.availableAbroad = availableAbroad;
    }
}

File: src/behavioral/visitor.ts
abstract class Client {
  constructor(
    protected name: string,
    protected address: string,
    protected number: string,
  ) {}
}

export class Resident extends Client {
  constructor(
    name: string,
    address: string,
    number: string,
    readonly insuranceClass: string,
  ) {
    super(name, address, number);
    this.insuranceClass = insuranceClass;
  }

  getName(): string {
    return this.name;
  }
}

export class Company extends Client {
  constructor(
    name: string,
    address: string,
    number: string,
    readonly nbrOfEmployees: number,
  ) {
    super(name, address, number);
    this.nbrOfEmployees = nbrOfEmployees;
  }
}

export class Bank extends Client {
  constructor(
    name: string,
    address: string,
    number: string,
    readonly branchesInsured: number,
  ) {
    super(name, address, number);
    this.branchesInsured = branchesInsured;
  }
}

export class Restaurant extends Client {
  constructor(
    name: string,
    address: string,
    number: string,
    readonly availableAbroad: boolean,
  ) {
    super(name, address, number);
    this.availableAbroad = availableAbroad;
  }
}
